File: q2594/q2594_bs.java
class Solution {
    public long repairCars(int[] ranks, int cars) {
        long lo = 1;
        long hi = 1L * ranks[0] * cars * cars;

        while (lo < hi) {
            long mid = (lo + hi) / 2;
            long totalCarsRepairedInMid = totalCarsRepaired(mid, ranks);
            if (totalCarsRepairedInMid < cars) {
                // If mid is not sufficient time to repair cars, increase lo
                // to find more time to repair all cars
                lo = mid + 1;
            } else {
                // If mid is sufficient time to repair cars, decrease hi to see
                // if we can repair the cars in less time
                hi = mid;
            }
        }
        return lo;
    }

    /** 
    * Compute the total number of cars that all mechanics can repair 
    * totalTime: The total time available for all mechanics
    * ranks: The ranks of the car mechanics
    */
    private long totalCarsRepaired(long totalTime, int[] ranks) {
        long totalCars = 0;
        for (int rank : ranks) {
            totalCars += (long)(Math.sqrt(totalTime / rank));
        }
        return totalCars;
    } 
}

File: q2594/q2594_bs.go
func repairCars(ranks []int, cars int) int64 {
	lo, hi := int64(1), int64(ranks[0]*cars*cars) // Initial low and high bounds for binary search

	// Binary search for the minimum time required to repair the cars
	for lo < hi {
		mid := (lo + hi) / 2
		totalCarsRepairedInMid := totalCarsRepaired(mid, ranks)

		if totalCarsRepairedInMid < int64(cars) {
			// If mid is not sufficient time to repair cars, increase lo
			// to find more time to repair all cars
			lo = mid + 1
		} else {
			// If mid is sufficient time to repair cars, decrease hi to see
			// if we can repair the cars in less time
			hi = mid
		}
	}
	return lo
}

func totalCarsRepaired(totalTime int64, ranks []int) int64 {
	var totalCars int64

	// Calculate total cars repaired for the given total time
	for _, rank := range ranks {
		totalCars += int64(math.Sqrt(float64(totalTime) / float64(rank)))
	}

	return totalCars
}

